### PR TITLE
Fix Tom's refactor of CLI, add support for U18.04 and Centos7.7 *merged against master

### PR DESCRIPTION
**Pull request by sarun87**
Fix Tom's large change so that the CLI:
    - works for Ubuntu 18.04
    - setup completes for both root & non-root users
    - add platform check and dependency installs
    - load active config before any command
    - remove bash auto-complete to speed up cli invocation
    - pf9ctl cluster prerp-node and bootstrap works for local & remote hosts
    - fixes for pf9ctl config creation / activation
    - fixes for pf9ctl cluster create
    - fixes for pf9ctl cluster attach-node
    - creates symlink in /usr/bin (using sudo) since bashrc is not sourced until next login
    - simplified user input question and logging text
    - generealize NTP ansible task
    - logging format used unicode "-". Replace this with ascii "-"
    - add support for centos* 7.6 uses py2.7
    - python3-distutils is u18 only. Separate out prereq for u18
    - install haveged to increase entropy. Low entropy causes pf9ctl to be slow/get stuck
    - Hide appcatalog flag, default to False
    - Fix get_resmgr_hostid to handle exceptions

    Testing:
    Manually tested prep-node, bootstrap, cluster create, attach-node on U18, Centos 7.6/7.7 and U16.

Merged master to branch to resolve merge commit erros.

### CODE REVIEW

File: cli-setup.sh [hunk 1]
```diff
@@ -322,39 +375,51 @@ pip_url="https://bootstrap.pypa.io/get-pip.py"
...
     if ! (ln -s ${cli_entrypoint} ${cli_exec} > /dev/null 2>&1); then
-	    assert "Failed to create Express-CLI symlink: ${cli_exec}"; fi
+	    assert "Failed to create Express CLI symlink: ${cli_exec}"; fi
 else
-    stdout_log "Express-CLI symlink already exist"
+    debugging "Express CLI symlink already exist"
 fi
+
+# Create symlink in /usr/bin
+if [ -L /usr/bin/pf9ctl ]; then
```
Comment (sarun87) [inline cli-setup.sh:414]: Possible need for doc changes around cleanup.

File: cli-setup.sh [hunk 2]
```diff
@@ -281,7 +260,76 @@ create_cli_config(){
     done
 }
 
+validate_platform() {
+  # check if running CentOS 7, Ubuntu 16.04, or Ubuntu 18.04
+  if [ -r /etc/centos-release ]; then
```
Comment (sarun87) [inline cli-setup.sh:265]: Possible to make this much simpler using /etc/os-release as used here:
https://github.com/platform9/pf9-du/blob/atherton/bbone/bbslave/service_scripts/pf9-service-functions.sh#L12

File: pf9/cluster/cluster_attach.py [hunk 3]
```diff
@@ -122,9 +122,8 @@ def get_uuids(self, host_ips):
             if host_uuid is not None:
                 host_uuids.append(host_uuid)
             else:
-                # TODO: We should warn the end user and proceed?
-                pass
-
+                logger.info("Unable to find host with IP:{}, please try again or run prep-node first".format(host_ip))
```
Comment (sarun87) [inline pf9/cluster/cluster_attach.py:125]: Return error code?

File: pf9/cluster/cluster_attach.py [hunk 4]
```diff
@@ -105,14 +105,14 @@ def get_resmgr_hostid(self, host_ip):
...
-                continue
-            # TODO: Why is there a loop in a loop where the values from the first loop aren't used?
-            for key, value in host['extensions']['interfaces']['data'].items():
+            try:
                 for iface_name, iface_ip in host['extensions']['interfaces']['data']['iface_ip'].items():
                     if iface_ip == host_ip:
                         logger.info("Node host_id: {}".format(host['id']))
                         return host['id']
+            except Exception as e:
+                logger.exception(e)
```
Comment (sarun87) [inline pf9/cluster/cluster_attach.py:114]: Return non-zero CLI exit code?

File: cli-setup.sh [hunk 5]
```diff
@@ -251,19 +230,19 @@ EOT
 }
 
 create_cli_config(){
-    debugging "Creating PF9 CLI Configuration"
+    debugging "Creating Platform9 Express CLI Configuration"
```
[the remark this answers is not in the document]
Comment (sarun87) [reply cli-setup.sh:233]: Will get rid of Express from all user facing o/p everywhere.

File: pf9/cluster/commands.py [hunk 6]
```diff
@@ -373,13 +387,17 @@ def attach_node(ctx, **kwargs):
               help='SSH key for nodes.')
 @click.option('--ips', '-i', multiple=True,
               help='IPs of the host to be prepared. Specify multiple IPs by repeating this option.')
-@click.option('--floating-ip', '-f', multiple=True, hidden=True)
+@click.option('--floating-ip', '-f', default=None, multiple=True, hidden=True)
```
[the remark this answers is not in the document]
Comment (sarun87) [reply pf9/cluster/commands.py:390]: This field is hidden so wont' be exposed to users.

It's being used by express scripts. Unclear how to use it though.

File: cli-setup.sh [hunk 7]
```diff
@@ -70,12 +70,11 @@ parse_args() {
...
-#	    echo "	  [--dev] Installs from local source code for each project in editable mode."
-#	    echo "                This assumes you have provided all source code in the correct locations"
-#	    echo "	  [--local] Installs local source code in the same directory"
-#	    echo "	  [-d|--debug]"
-#	    echo ""
+	    echo "	  [--branch=] Specify a different branch to pull Platform9 Express CLI source"
+	    echo "	  [--dev] Installs from local source code for each project in editable mode."
+	    echo "                This assumes you have provided all source code in the correct locations"
+	    echo "	  [--local] Installs local source code in the same directory"
+	    echo "	  [-d|--debug]"
```
[the remark this answers is not in the document]
Comment (sarun87) [reply cli-setup.sh]: Fixed.

File: cli-setup.sh (hunk 2 above)
Comment (sarun87) [reply cli-setup.sh:265]: Punting this for now. Needs more testing across U16, U18, Centos7.6 and 7.7.

File: pf9/cluster/cluster_attach.py (hunk 4 above)
Comment (sarun87) [reply pf9/cluster/cluster_attach.py:114]: There is no exit code handling in the code base. This needs a larger change. I suggest we take this up later if need be. The exception is still logged to the log file. That will provide debug information.

File: cli-setup.sh (hunk 2 above)
Comment (sarun87) [reply cli-setup.sh:265]: Taking this in since more changes are being tested.

File: cli-setup.sh [hunk 8]
```diff
@@ -262,26 +285,22 @@ create_cli_config(){
...
+  elif [[ "$ID" == "centos" ]]; then
+     platform="centos"
+     if [[ "$VERSION_ID" != "7" ]]; then
+         stdout_log "Unsupported Centos version"
+         exit 99
+     fi
   else
-    stdout_log "Unsupported platform"
-    exit 99
+     stdout_log "Unsupported linux distribution"
```
Comment (pshanbhag) [inline cli-setup.sh:302]: nit: Linux

File: cli-setup.sh [hunk 9]
```diff
@@ -262,26 +285,22 @@ create_cli_config(){
...
+  if [[ "$ID" == "ubuntu" ]]; then
+     platform="ubuntu"
+     if [[ "$VERSION_ID" != "18.04" ]] && [[ "$VERSION_ID" != "16.04" ]]; then
+         stdout_log "Unsupported Ubuntu version"
+         exit 99
+     fi
+  elif [[ "$ID" == "centos" ]]; then
+     platform="centos"
+     if [[ "$VERSION_ID" != "7" ]]; then
+         stdout_log "Unsupported Centos version"
```
Comment (pshanbhag) [inline cli-setup.sh:298]: nit: CentOS

File: cli-setup.sh [hunk 10]
```diff
@@ -235,12 +253,17 @@ create_cli_config(){
...
 	    stdout_log "Attempt $auth_retry of ${max_auth_retry}:"; fi
-	eval "${cli_exec}" config create
-	
+
+    if [ ! -z "${PF9_MGMTURL}" ] && [ ! -z "${PF9_USER}" ] && [ ! -z "${PF9_PASS}" ] && [ ! -z "${PF9_REGION}" ] && [ ! -z "${PF9_TENANT}" ]; then
+        eval "${cli_exec}" config create --du_url ${PF9_MGMTURL} --os_username ${PF9_USER} --os_password ${PF9_PASS} --os_region ${PF9_REGION} --os_tenant ${PF9_TENANT}
+    else
+        echo ""
+        stdout_log "Please provide your Platform9 Credentials"
+        eval "${cli_exec}" config create
```
Comment (pshanbhag) [inline cli-setup.sh:264]: Is the tenant name and region mandatory now? If so, this is a change in previous behavior. @chrisj-platform9 needs to update the documentation around it.

File: cli-setup.sh (hunk 8 above)
Comment (sarun87) [reply cli-setup.sh:302]: fixed

File: cli-setup.sh (hunk 9 above)
Comment (sarun87) [reply cli-setup.sh:298]: fixed

File: cli-setup.sh (hunk 10 above)
Comment (sarun87) [reply cli-setup.sh:264]: Interactive experience for the `config create` command has the functionality to default `project` and `region` for FT users but it isn't there as part of `cli-setup.sh` after Tom's changes.

After discussion with @chrisj-platform9  , the description below is added as part of Interactive experience for the `config create` command:
```Enter Region & Tenant Details (Freedom Plan requires RegionOne and service)```

Documentation in general needs update for the modified UX.

**Pull request merged**